src/if_statements.py

```diff
@@ -353,55 +288,208 @@ def immediate_postdominator(context: Context, start: Node, end: Node) -> Node:
+    conditions: List[Condition] = []
+    bottom = start.conditional_edge
+    curr_node: ConditionalNode = start
+    for num_conditions in itertools.count(1):
+        # Collect conditions as we accumulate them:
+        block_info = curr_node.block.block_info
+        assert isinstance(block_info, BlockInfo)
+        branch_condition = block_info.branch_condition
+        assert branch_condition is not None
+        if num_conditions == 1:
+            # The first condition in an if-statement will have
+            # unrelated statements in its to_write list. Circumvent
+            # emitting them twice by just using branch_condition:
+            conditions.append(branch_condition)
+        # Make sure to write down each block's statement list,
+        # even inside an and/or group.
+        else:
+            conditions.append(get_full_condition(block_info))
+
+        # The next node will tell us whether we are in an &&/|| statement...
+        next_node = curr_node.fallthrough_edge
+
+        else_body: Optional[Body]
+        if is_and_statement(next_node, bottom):
+            # We reached the end of an and-statement.
+            # TODO: The last one - or more - might've been part
+            # of a while loop.
+
+            else_body = build_flowgraph_between(context, bottom, end, indent + 4)
+            if not any(stmt.should_write() for stmt in else_body.statements):
+                else_body = None
+            return IfElseStatement(
+                # We negate everything, because the conditional edges will jump
+                # OVER the if body.
+                join_conditions(conditions, "&&", only_negate_last=False),
+                indent,
+                context.options.coding_style,
+                if_body=build_flowgraph_between(context, next_node, end, indent + 4),
+                else_body=else_body,
+            )
+        assert isinstance(next_node, ConditionalNode)  # from is_and_statement()
+
+        if next_node.fallthrough_edge is bottom:
+            assert next_node.block.block_info
+            next_node_condition = next_node.block.block_info.branch_condition
+            assert next_node_condition
+            # We were following an or-statement.
+            return IfElseStatement(
+                # Negate the last condition, for it must fall-through to the
+                # body instead of jumping to it, hence it must jump OVER the body.
+                join_conditions(
+                    conditions + [next_node_condition], "||", only_negate_last=True
+                ),
+                indent,
+                context.options.coding_style,
+                if_body=build_flowgraph_between(context, bottom, end, indent + 4),
+                # The else-body is wherever the code jumps to instead of the
+                # fallthrough (i.e. if-body).
+                else_body=build_flowgraph_between(
+                    context, next_node.conditional_edge, end, indent + 4
+                ),
```

Comment:
the order in which build_flowgraph_between is called matters in rare cases, because of goto emission. Use local variables for the results of the two calls. Also, this else_body should sometimes be None, just like for &&